runtime/cudaq/operators/sum_op.cpp

```diff
@@ -546,6 +546,39 @@ INSTANTIATE_SUM_EVALUATIONS(boson_handler);
 INSTANTIATE_SUM_EVALUATIONS(fermion_handler);
 #endif
 
+// Adjoint
+
+template <typename HandlerTy>
+sum_op<HandlerTy> sum_op<HandlerTy>::adjoint() const {
```

Comment:
You could make this more efficient by special casing for operator classes that can be canonicalized. If they can be canonicalized, then the operator framework guarantees that they are (you can `assert(assert(!HandlerTy::can_be_canonicalized || this->is_canonicalized());)`). In the case of a canonicalized operator, each degree is acted on only once. Regardless of commutation relation, this means that adjointing preserves the order of operators (not need to product multiply, you can just create a vector and populate it with the adjoint of the original in the same order).